src/de.rs

```diff
@@ -895,3 +1150,167 @@ pub async fn read_from<S: AsyncReadExt + Send + Unpin, T: FromStream>(
+
+    use super::*;
+
+    /// next_or_eof should return the next char in the buffer/stream, or
+    /// if we've hit the EOF, throw an error.
+    #[tokio::test]
+    async fn test_next_or_eof() {
+        let s = b"bar";
+        for num_chunks in (1..s.len()).rev() {
+            let source = stream::iter(s.iter().copied())
+                .chunks(num_chunks)
+                .map(Bytes::from)
+                .map(Result::<Bytes, Error>::Ok);
+
+            let mut decoder = Decoder::from_stream(source);
+            for expected in s {
+                let actual = decoder.next_or_eof().await.unwrap();
+                assert_eq!(&actual, expected);
+            }
+            let res = decoder.next_or_eof().await;
+            assert!(res.is_err());
+        }
+    }
+
+    /// ignore_exactly takes a vector of bytes, and consumes exactly those characters.
+    #[tokio::test]
+    async fn test_ignore_exactly() {
+        let inputs = [("foo", "foo", true, 0), ("foobar", "foo", true, 3)];
+
+        for input in inputs {
+            let source = input.0;
+            let to_ignore = input.1;
+            let success = input.2;
+            let chars_left = input.3;
+
+            let source = stream::iter(source.as_bytes().iter().copied())
+                .chunks(source.len())
+                .map(Bytes::from)
+                .map(Result::<Bytes, Error>::Ok);
+
+            let mut decoder = Decoder::from_stream(source);
+
+            let res = decoder.ignore_exactly(to_ignore).await;
+            assert_eq!(res.is_ok(), success);
+            assert_eq!(decoder.buffer.len(), chars_left);
+        }
+    }
+
+    /// ignore_string should consume the stream until the end of the current string.
+    #[tokio::test]
+    async fn test_ignore_string() {
+        let inputs = [
+            ("\"foo\"bar", 3),
+            ("\"test\"", 0),
+            ("\"\"", 0),
+            ("\"\\r\"", 0),
+            ("\"hello\"world\"", 6),
+            ("\"   hello\"", 0),
+            ("\"hello   \"   ", 3),
+            ("\"\\t\\n\\r\"", 0),
+            ("\"\"test\\\"", 6),
+            ("\"\\\\\\\\\"", 0),
+        ];
+
+        for input in inputs {
+            let source = input.0;
+            let end_length = input.1;
+
+            let source = stream::iter(source.as_bytes().iter().copied())
+                .chunks(source.len())
+                .map(Bytes::from)
+                .map(Result::<Bytes, Error>::Ok);
+
+            let mut decoder = Decoder::from_stream(source);
+
+            decoder.ignore_string().await.unwrap();
+            assert_eq!(decoder.buffer.len(), end_length);
+        }
+    }
+
+    #[tokio::test]
+    async fn test_ignore_number() {
+        let inputs = [
+            ("0", 0),
+            ("123", 0),
+            ("-123", 0),
+            ("123.45", 0),
+            ("-123.45", 0),
+            ("0.0", 0),
+            ("123, 45", 4),
+        ];
+
+        for input in inputs {
```

Comment:
should probably move these into `#[test_case(...)]`s like the following two tests